drivers/adc/ad7606/ad7606.c

```diff
@@ -600,6 +602,71 @@ int32_t ad7606_reg_write_mask(struct ad7606_dev *dev,
 	return ad7606_reg_write(dev, addr, reg_data);
 }
 
+/***************************************************************************
+ * @brief Correction bit31 - bit18 for negative value (for 18 bits ADC7606 data via spi)
+ *
+ * @param dev       - The device structure pointer
+ * @param pdata     - data buffer pointer
+ * @param size      - Number of data needs correct
+ * @return 
+ *******************************************************************************/
+static void adcdata_sign_correction(struct ad7606_dev* dev, uint32_t size, uint32_t* pdata)
+{
+
+	uint8_t i = 0;
+
+    //correct upper 14 bits for adc data
+    for(i = 0;i < size;i ++)
+    {
+        if((dev->range_ch[i].min) < 0)   //if negative value could exist
+        {
+            if( * pdata >> 17)            //if sign bit is 1(negative)
+            {
+                * pdata = * pdata | ((uint32_t)0xFFFC << 16); //set bit31 - bit18 to 1
+            }
```

Comment:
I think you're trying to sign extend here, right? We already have `no_os_sign_extend()`.